File: overlay/python/hydra-kernel/src/main/python/hydra/lib/optionals.py
"""Python implementations of hydra.lib.optionals primitives."""

from __future__ import annotations
from collections.abc import Callable, Sequence
from typing import Any, TypeVar
from hydra.dsl.python import frozenlist, Optional, Given, None_, NONE_

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')


def apply(f: Optional[Callable[[A], B]], x: Optional[A]) -> Optional[B]:
    """Apply a function to an argument (applicative)."""
    match f:
        case None_():
            return NONE_
        case Given(func):
            match x:
                case None_():
                    return NONE_
                case Given(val):
                    return Given(func(val))


def bind(x: Optional[A], f: Callable[[A], Optional[B]]) -> Optional[B]:
    """Chain operations on optional values, handling Nothing cases automatically."""
    match x:
        case None_():
            return NONE_
        case Given(val):
            return f(val)


def cases(m: Optional[A], n: B | Callable[[], B], j: Callable[[A], B]) -> B:
    """Handle an optional value with the maybe value as the first argument."""
    match m:
        case None_():
            return n() if callable(n) else n  # type: ignore[return-value]
        case Given(val):
            return j(val)


def cat(xs: Sequence[Optional[A]]) -> frozenlist[A]:
    """Filter out Nothing values from a list."""
    result: list[A] = []
    for x in xs:
        match x:
            case Given(val):
                result.append(val)
            case None_():
                pass
    return tuple(result)


def compose(f: Callable[[A], Optional[B]], g: Callable[[B], Optional[C]], x: A) -> Optional[C]:
    """Compose two Maybe-returning functions (Kleisli composition)."""
    return bind(f(x), g)


def from_optional(default: A | Callable[[], A], x: Optional[A]) -> A:
    """Get a value from an optional value, or return a default value."""
    match x:
        case Given(val):
            return val
        case None_():
            return default() if callable(default) else default  # type: ignore[return-value]


def is_given(x: Optional[Any]) -> bool:
    """Check if a value is Just."""
    return isinstance(x, Given)


def is_none(x: Optional[Any]) -> bool:
    """Check if a value is Nothing."""
    return isinstance(x, None_)


def map(f: Callable[[A], B], x: Optional[A]) -> Optional[B]:
    """Map a function over an optional value."""
    match x:
        case Given(val):
            return Given(f(val))
        case None_():
            return NONE_


def map_optional(f: Callable[[A], Optional[B]], xs: Sequence[A]) -> frozenlist[B]:
    """Map a function over a list and collect Just results."""
    result: list[B] = []
    for x in xs:
        y = f(x)
        match y:
            case Given(val):
                result.append(val)
            case None_():
                pass
    return tuple(result)


def pure(x: A) -> Optional[A]:
    """Lift a value into the Maybe type."""
    return Given(x)


def to_list(x: Optional[A]) -> frozenlist[A]:
    """Convert a Maybe to a list: Just x becomes [x], Nothing becomes []."""
    match x:
        case Given(val):
            return (val,)
        case None_():
            return ()

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Haskell/Lib/Optionals.hs
-- | Haskell implementations of hydra.lib.optionals primitives

module Hydra.Haskell.Lib.Optionals where

import qualified Data.Maybe as Y


-- | Apply a function to an argument (applicative).
apply :: Y.Maybe (a -> b) -> Y.Maybe a -> Y.Maybe b
apply = (<*>)

-- | Chain operations on optional values, handling absent cases automatically.
bind :: Y.Maybe a -> (a -> Y.Maybe b) -> Y.Maybe b
bind = (>>=)

-- | Handle an optional value with the absent-case value as the first argument.
cases :: Y.Maybe a -> b -> (a -> b) -> b
cases m n j = Y.maybe n j m

-- | Filter out absent values from a list.
cat :: [Y.Maybe a] -> [a]
cat = Y.catMaybes

-- | Compose two optional-returning functions (Kleisli composition).
compose :: (a -> Y.Maybe b) -> (b -> Y.Maybe c) -> (a -> Y.Maybe c)
compose f g = \x -> f x >>= g

-- | Get a value from an optional value, or return a default value.
fromOptional :: a -> Y.Maybe a -> a
fromOptional = Y.fromMaybe

-- | Check if a value is present.
isGiven :: Y.Maybe a -> Bool
isGiven = Y.isJust

-- | Check if a value is absent.
isNone :: Y.Maybe a -> Bool
isNone = Y.isNothing

-- | Map a function over an optional value.
map :: (a -> b) -> Y.Maybe a -> Y.Maybe b
map = fmap

-- | Map a function over a list and collect present results.
mapOptional :: (a -> Y.Maybe b) -> [a] -> [b]
mapOptional = Y.mapMaybe

-- | Lift a value into the optional type.
pure :: a -> Y.Maybe a
pure = Just

-- | Convert an optional value to a list: a present value becomes [x], an absent value becomes [].
toList :: Y.Maybe a -> [a]
toList = Y.maybeToList
